File: 200. Number of Islands/solution.py
class Solution:
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        # edge case
        if not grid or not grid[0]:
            return 0
        
        # count by dfs
        count = 0
        n = len(grid)
        m = len(grid[0])
        for i in range(n):
            for j in range(m):
                if grid[i][j] == "1":
                    self.dfs(grid, i, j)
                    count += 1
        return count
                    
    def dfs(self, grid, i, j):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] != "1":
            return
        
        grid[i][j] = "-1"
        self.dfs(grid, i-1, j)
        self.dfs(grid, i, j-1)
        self.dfs(grid, i, j+1)
        self.dfs(grid, i+1, j)

File: 200. Number of Islands/solution.java
class Solution {
    public int numIslands(char[][] grid) {
        // edge cases
        int m = grid.length;
        if (m == 0) {
            return 0;
        }
        
        int n = grid[0].length;
        if (n == 0) {
            return 0;
        }
        
        // count islands using dfs
        int count = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    dfs(grid, i, j, m, n);
                    count++;
                }
            }
        }
        
        return count;
    }
    
    public void dfs(char[][] grid, int i, int j, int m, int n) {
        if (i >= 0 && i < m && j >= 0 && j < n && grid[i][j] == '1') {
            grid[i][j] = '#';
            dfs(grid, i-1, j, m, n);
            dfs(grid, i, j-1, m, n);
            dfs(grid, i, j+1, m, n);
            dfs(grid, i+1, j, m, n);
        }
    }
}
